### PR TITLE
feat(auth): map OpenID groups claim

### LINKED ISSUE #2523: [Feat]: Allow specifying user and group scopes with OIDC authentication
### Is your feature request related to a problem? Please describe.

It should be possible to specify which claim to use for the user ID. Some users might want to use `email`, others might want to use `sub`.
Additionally, it should be possible to select a claim which is used for groups. This is already possible with LDAP: https://github.com/project-zot/zot/blob/002ff05f6e5ab3870aca75e616e8a4ed9b4a5c68/pkg/api/config/config.go#L165

### Describe the solution you'd like

Add `userclaim` and `groupsclaim` to OIDC provider config

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AkashKumar7902**
## What

Adds `claimMapping.groups` for OpenID providers so deployments can map group membership from a custom OIDC claim. The existing `claimMapping.username` behavior remains unchanged, and group mapping defaults to `groups` when unset.

Updates the OpenID claim-mapping example and README snippet.

Fixes #2523

## Validation

- `GOEXPERIMENT=jsonv2 go test -tags 'sync scrub metrics search lint mgmt' ./pkg/api -run TestRoutes -count=1`\n- `GOEXPERIMENT=jsonv2 go test -tags 'sync scrub metrics search lint mgmt' ./pkg/api -run 'TestRoutes|TestAuthn' -count=1`\n- `GOEXPERIMENT=jsonv2 go test ./pkg/api/config -run 'TestConfig|TestUpdateReloadableConfig|TestSanitize' -count=1`\n- `GOEXPERIMENT=jsonv2 go test ./pkg/cli/server -run 'TestSchema|TestVerify|TestLoadConfig' -count=1`\n- `jq empty examples/config-openid-claim-mapping.json`\n- `git diff --check`

### CODE REVIEW

File: pkg/api/routes.go [hunk 1]
```diff
@@ -2129,50 +2193,21 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallbackWithProvider(providerName stri
...
+		if username != "" {
 			rh.c.Log.Debug().
 				Str("provider", providerName).
+				Str("claim", usernameClaim).
 				Str("username", username).
-				Msg("using email as username (fallback)")
+				Msg("extracted username from configured claim")
 		}
 
 		if username == "" {
```
Comment (andaaron) [inline pkg/api/routes.go]: Let's use an if/else here. The username value is also checked above.

File: pkg/api/routes.go [hunk 2]
```diff
@@ -2184,22 +2219,16 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallbackWithProvider(providerName stri
...
 		var groups []string
 
-		val, ok := info.Claims["groups"].([]any)
-		if !ok {
-			rh.c.Log.Info().Msgf("failed to find any 'groups' claim for user %s in UserInfo", username)
-		}
-
-		for _, group := range val {
-			groups = append(groups, fmt.Sprint(group))
+		userInfoGroupCount := len(groups)
```
Comment (andaaron) [inline pkg/api/routes.go]: But groups is always 0 here, correct?

File: pkg/api/routes.go [hunk 3]
```diff
@@ -2184,22 +2219,16 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallbackWithProvider(providerName stri
...
+		groups = appendOpenIDGroups(groups, info.Claims, groupsClaim)
+		if len(groups) == userInfoGroupCount {
```
Comment (andaaron) [inline pkg/api/routes.go]:
```suggestion
		if len(groups) == 0 {
```

File: pkg/api/routes.go (hunk 1 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Done — collapsed into a nested if so the empty-username check happens in one place.

File: pkg/api/routes.go (hunk 2 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Right — replaced the length-based detection with a bool from `appendOpenIDGroups` in the followup commit, so we no longer rely on the saved count.

File: pkg/api/routes.go (hunk 3 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Switched to having `appendOpenIDGroups` return a found bool, so the length comparison is gone entirely.

File: pkg/api/routes.go [hunk 4]
```diff
@@ -2281,24 +2319,30 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallbackWithProvider(providerName stri
...
+			rh.c.Log.Info().Msgf("failed to find any %q claim for user %s in UserInfo", groupsClaim, username)
 		}
 
-		for _, group := range val {
-			groups = append(groups, fmt.Sprint(group))
+		idTokenGroupCount := len(groups)
+		groups = appendOpenIDGroups(groups, tokens.IDTokenClaims.Claims, groupsClaim)
+		if len(groups) == idTokenGroupCount {
+			rh.c.Log.Info().Msgf("failed to find any %q claim for user %s in IDTokenClaimsToken", groupsClaim, username)
 		}
```
[the remark this answers is not in the document]
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Good catch — `appendOpenIDGroups` now returns a found bool that distinguishes "claim missing/wrong type" from "claim present but empty", and the log uses that instead of length.

File: pkg/api/routes.go [hunk 5]
```diff
@@ -2281,24 +2319,30 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallbackWithProvider(providerName stri
...
-		if !ok {
-			rh.c.Log.Info().Msgf("failed to find any 'groups' claim for user %s in UserInfo", username)
+		if usernameConfigured {
+			rh.c.Log.Debug().
+				Str("provider", providerName).
+				Str("claim", usernameClaim).
+				Str("username", username).
+				Msg("extracted username from configured claim")
+		} else {
+			rh.c.Log.Info().
```
[the remark this answers is not in the document]
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Agreed, dropped the fallback log back to Debug.

File: pkg/api/routes.go [hunk 6]
```diff
@@ -2217,6 +2217,86 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallback() rp.CodeExchangeUserinfoCall
...
 
+const (
+	defaultUsernameClaim = "email"
+	defaultGroupsClaim   = "groups"
+)
+
+// getOpenIDClaimMapping resolves the username and groups claim names for a given provider.
+// The third return value reports whether the username claim was explicitly configured
+// (false means the default "email" claim is being used as a fallback).
+func getOpenIDClaimMapping(authConfig *config.AuthConfig, providerName string) (string, string, bool) {
```
Comment (andaaron) [inline pkg/api/routes.go]: Maybe we should move this to authn.go. Its only relevant for authentication and doesn't depend on anything else here.

File: pkg/api/routes.go [hunk 7]
```diff
@@ -2217,6 +2217,86 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallback() rp.CodeExchangeUserinfoCall
...
+	}
+
+	if providerConfig.ClaimMapping.Groups != "" {
+		groupsClaim = providerConfig.ClaimMapping.Groups
+	}
+
+	return usernameClaim, groupsClaim, usernameConfigured
+}
+
+func getOpenIDUsername(info *oidc.UserInfo, claimName string) string {
```
Comment (andaaron) [inline pkg/api/routes.go]: Maybe we should move this to authn.go. Its only relevant for authentication and doesn't depend on anything else here.

File: pkg/api/routes.go [hunk 8]
```diff
@@ -2228,77 +2308,50 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallbackWithProvider(providerName stri
 		relyingParty rp.RelyingParty, info *oidc.UserInfo,
 	) {
 		// Extract username based on claim mapping configuration
-		var username string
 		authConfig := rh.c.Config.CopyAuthConfig()
```
Comment (andaaron) [inline pkg/api/routes.go:2528]: Maybe we should move this whole block up until line 2358 into a new function in authn.go

File: pkg/api/routes.go [hunk 9]
```diff
@@ -2217,6 +2217,86 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallback() rp.CodeExchangeUserinfoCall
...
+	default:
+		if val, ok := info.Claims[claimName].(string); ok {
+			return val
+		}
+	}
+
+	return ""
+}
+
+func appendOpenIDGroups(groups []string, claims map[string]any, claimName string) ([]string, bool) {
```
Comment (andaaron) [inline pkg/api/routes.go]: Maybe we should move this to authn.go. Its only relevant for authentication and doesn't depend on anything else here.

File: pkg/api/openid_claim_mapping_internal_test.go [hunk 10]
```diff
@@ -0,0 +1,266 @@
+//go:build sync && scrub && metrics && search && lint && mgmt
```
Comment (andaaron) [inline pkg/api/openid_claim_mapping_internal_test.go]: These labels shouldn't be needed.

File: pkg/api/routes.go [hunk 11]
```diff
@@ -2228,77 +2308,50 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallbackWithProvider(providerName stri
...
-							Str("claim", claimName).
-							Str("username", username).
-							Msg("extracted username from configured claim")
-					}
-				}
+		username := getOpenIDUsername(info, usernameClaim)
+		if username == "" {
+			if usernameClaim != defaultUsernameClaim {
+				usernameClaim = defaultUsernameClaim
+				usernameConfigured = false
```
Comment (andaaron) [inline pkg/api/routes.go]: When an operator configures claimMapping.username = "preferred_username" (or any custom claim) and the IdP returns it empty/absent, we silently reset usernameConfigured = false and log "using email as username (fallback)" at Debug — losing the signal that the configured claim was tried and missing.

File: pkg/api/routes.go [hunk 12]
```diff
@@ -2217,6 +2217,86 @@ func (rh *RouteHandler) OpenIDCodeExchangeCallback() rp.CodeExchangeUserinfoCall
...
+			return val
+		}
+	}
+
+	return ""
+}
+
+func appendOpenIDGroups(groups []string, claims map[string]any, claimName string) ([]string, bool) {
+	switch val := claims[claimName].(type) {
+	case []any:
```
Comment (andaaron) [inline pkg/api/routes.go]: The string branch filters empty values, but []string and []any don't. With []any{"dev", nil, ""} you'd end up with "<nil>" and "" group entries (and they'd survive slices.Compact). Three options:

Filter empties / skip nil entries consistently in all three branches.
Or document the intent (e.g., "we trust the IdP to send only non-empty entries").
Adding a test case like {"roles": []any{nil, ""}} would also pin the chosen behavior down.

File: pkg/api/authn_internal_test.go
Comment (andaaron) [inline pkg/api/authn_internal_test.go:1]: Rename to authn_internal_test.go

**Comment by andaaron** [comment]
Please rebase this PR

File: pkg/api/routes.go (hunk 6 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Done — moved `getOpenIDClaimMapping` (along with the `defaultUsernameClaim`/`defaultGroupsClaim` constants) into authn.go in 30160a72.

File: pkg/api/routes.go (hunk 7 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Moved `getOpenIDUsername` over to authn.go as well.

File: pkg/api/routes.go (hunk 9 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Same — `appendOpenIDGroups` is now in authn.go.

File: pkg/api/routes.go (hunk 8 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go:2528]: Good idea — pulled the whole claim/group extraction block into a new `extractOpenIDIdentity` helper in authn.go. The HTTP callback in routes.go is now just `extractOpenIDIdentity(...) -> OAuth2Callback(...)` plus error handling, which keeps it readable and lets the helper be unit-tested directly.

File: pkg/api/openid_claim_mapping_internal_test.go (hunk 10 above)
Comment (AkashKumar7902) [reply pkg/api/openid_claim_mapping_internal_test.go]: Removed — the file no longer has any build tags.

File: pkg/api/authn_internal_test.go
Comment (AkashKumar7902) [reply pkg/api/authn_internal_test.go:1]: Done — renamed to `authn_internal_test.go` (via `git mv` so the history is preserved).

File: pkg/api/routes.go (hunk 11 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Good point — silently resetting and only logging at Debug hid the signal. Now `extractOpenIDIdentity` logs a Warn (`configured username claim missing or empty, falling back to email`) including the configured claim name before falling back, so an operator can see exactly what they configured and why it didn't match.

File: pkg/api/routes.go (hunk 12 above)
Comment (AkashKumar7902) [reply pkg/api/routes.go]: Went with option 1 — `appendOpenIDGroups` now filters `nil` and empty entries consistently across all three branches (`[]any`, `[]string`, and `string`). Added two test cases pinning the behavior down: `[]any{"dev", nil, ""}` → `["dev"]` and `[]string{"admin","","ops"}` → `["admin","ops"]`.

**Comment by AkashKumar7902** [comment]
Rebased onto main in 30160a72. The rebase commit also folds in the round of changes from your latest review (helpers and the extraction block moved into authn.go behind a single `extractOpenIDIdentity` helper, build tags dropped, test file renamed to `authn_internal_test.go`, group filter consistent across all branches, and a Warn log when a configured username claim is missing/empty).

File: pkg/api/config/config.go
Comment (andaaron) [inline pkg/api/config/config.go:319]: Either remove this line or update it to specify the user groups

File: pkg/api/authn.go [hunk 13]
```diff
@@ -1106,6 +1106,161 @@ func saveUserLoggedSession(cookieStore sessions.Store, response http.ResponseWri
...
+
+	if idTokenClaims != nil {
+		groups, groupsFound = appendOpenIDGroups(groups, idTokenClaims, groupsClaim)
+		if !groupsFound {
+			logger.Info().Msgf("failed to find any %q claim for user %s in IDTokenClaimsToken", groupsClaim, username)
+		}
+	}
+
+	slices.Sort(groups)
+	groups = slices.Compact(groups)
```
Comment (andaaron) [inline pkg/api/authn.go:1259]: Since it is likely groups in info and idTokenClaims aren't both populated for every request, this would probably result in one of the messages at line 1247 and 1254 always be emitted, resulting in a lot of log messages.

I suggest we only log if the groups slice is empty here, a debug log level message.

File: pkg/api/authn.go [hunk 14]
```diff
@@ -1106,6 +1106,161 @@ func saveUserLoggedSession(cookieStore sessions.Store, response http.ResponseWri
...
+	if username == "" && usernameConfigured {
+		configuredClaim := usernameClaim
+		usernameClaim = defaultUsernameClaim
+		usernameConfigured = false
+		username = getOpenIDUsername(info, usernameClaim)
+
+		logger.Warn().
+			Str("provider", providerName).
+			Str("claim", configuredClaim).
+			Msg("configured username claim missing or empty, falling back to email")
```
Comment (andaaron) [inline pkg/api/authn.go:1219]: If an operator explicitly sets claimMapping.username = "email" (matching the default) and the email is missing, this branch fires (usernameConfigured is true) and re-queries the same defaultUsernameClaim, so the second getOpenIDUsername call returns the same empty string and the message reads "configured username claim 'email' missing or empty, falling back to email". Either:

Skip the warn/refallback when configuredClaim == defaultUsernameClaim (the re-query is a no-op anyway), or
Rephrase the message to "configured username claim missing or empty" without the "falling back to email" tail when the configured claim already is email.

File: pkg/api/authn.go [hunk 15]
```diff
@@ -1106,6 +1106,161 @@ func saveUserLoggedSession(cookieStore sessions.Store, response http.ResponseWri
...
+	}
+
+	slices.Sort(groups)
+	groups = slices.Compact(groups)
+
+	return username, groups, true
+}
+
 // OAuth2Callback is the callback logic where openid/oauth2 will redirect back to our app.
 func OAuth2Callback(ctlr *Controller, w http.ResponseWriter, r *http.Request, state, email, provider string,
```
Comment (andaaron) [inline pkg/api/authn.go:1265]: Can we rename "email" to "username" for consistency?

**Review by andaaron — APPROVED** [review]
Will approve this for now and take care of the nits in a separate PR.

**Comment by andaaron** [comment]
> Will approve this for now and take care of the nits in a separate PR.

https://github.com/project-zot/zot/pull/4028

File: pkg/api/authn.go [hunk 16]
```diff
@@ -1106,6 +1106,161 @@ func saveUserLoggedSession(cookieStore sessions.Store, response http.ResponseWri
...
+func extractOpenIDIdentity(logger log.Logger, authConfig *config.AuthConfig, providerName string,
+	info *oidc.UserInfo, idTokenClaims map[string]any,
+) (string, []string, bool) {
+	usernameClaim, groupsClaim, usernameConfigured := getOpenIDClaimMapping(authConfig, providerName)
+
+	username := getOpenIDUsername(info, usernameClaim)
+
+	if username == "" && usernameConfigured {
+		configuredClaim := usernameClaim
+		usernameClaim = defaultUsernameClaim
```
Comment (vrajashkr) [inline pkg/api/authn.go:1212]: IMO, if the user has explicitly configured a claimMapping, we probably shouldn't try to fall back to email during claim evaluation.

Using email as a default claim is fine if the user has not configured anything.

File: pkg/api/authn.go [hunk 17]
```diff
@@ -1106,6 +1106,161 @@ func saveUserLoggedSession(cookieStore sessions.Store, response http.ResponseWri
...
+			Str("provider", providerName).
+			Str("claim", usernameClaim).
+			Str("username", username).
+			Msg("extracted username from configured claim")
+	} else {
+		logger.Debug().
+			Str("provider", providerName).
+			Str("username", username).
+			Msg("using email as username (fallback)")
+	}
```
Comment (vrajashkr) [inline pkg/api/authn.go:1237]: These debug logs seem to be based on `usernameConfigured` which if I'm not mistaken is deterministic based on the config that the user has provided. As such, it may not be a helpful log message.

Instead, a generic log mentioning the provider, claim, and username in all cases might be better. That said, there is also a question as to whether usernames should be logged.

File: pkg/api/authn.go [hunk 18]
```diff
@@ -1106,6 +1106,161 @@ func saveUserLoggedSession(cookieStore sessions.Store, response http.ResponseWri
...
+
+	var (
+		groups      []string
+		groupsFound bool
+	)
+
+	if info != nil {
+		groups, groupsFound = appendOpenIDGroups(groups, info.Claims, groupsClaim)
+		if !groupsFound {
+			logger.Info().Msgf("failed to find any %q claim for user %s in UserInfo", groupsClaim, username)
```
Comment (vrajashkr) [inline pkg/api/authn.go:1247]: Having this in info would make the logs quite a bit chatty.
Would recommend moving it to debug.

Also, "failed" is often used in context of errors which this is not. Would instead suggest something like below:


```suggestion
			logger.Info().Msgf("no groups found for user %s in UserInfo", groupsClaim, username)
```

File: pkg/api/authn.go [hunk 19]
```diff
@@ -1106,6 +1106,161 @@ func saveUserLoggedSession(cookieStore sessions.Store, response http.ResponseWri
...
+		}
+	}
+
+	if idTokenClaims != nil {
+		groups, groupsFound = appendOpenIDGroups(groups, idTokenClaims, groupsClaim)
+		if !groupsFound {
+			logger.Info().Msgf("failed to find any %q claim for user %s in IDTokenClaimsToken", groupsClaim, username)
```
Comment (vrajashkr) [inline pkg/api/authn.go:1254]: Similar comment as above.

File: pkg/api/authn.go (hunk 16 above)
Comment (andaaron) [reply pkg/api/authn.go:1212]: This was also the case before this PR https://github.com/project-zot/zot/pull/3999/changes#diff-20343ca31da2491d1160ba5ed434f5dfa0698341d7ed516306872ebe03758b49L2569

Let's not change behavior.

**Comment by andaaron** [comment]
@vrajashkr I opened https://github.com/project-zot/zot/pull/4028 to tackle most of these aspects. Please check it out
